### DISCUSSION TITLE
How to get sub-image or reuse image

### DISCUSSION THREAD

**Discussion by andrea-gamedev** [fltk-rs/fltk-rs#937] [Q&A] [answered]
I'm trying to make a sprite editing program that I'd like to allow importing of sprite sheets and slicing into sprites within the program for export to other formats. I'm trying to display the whole spritesheet unchanged as one image, but I'd also like to have another section of the UI that shows a scrolling list of the sprites you've marked out in that spritesheet, without having to store the sprites on disk as new images. I've thought of a few possible ways to do it, but I'm unsure if any of them is achievable in fltk-rs, and if so how to actually achieve them

1) Sub-Images/Sprites: only the original image is loaded into memory, and the list of sprites somehow just renders a chunk of pixels from that original image without creating any new images
2) New images in memory: When a sprite is marked on the sprite sheet, a new image is created in memory, copying a rectangle of pixels from the original image. This seems like the most likely solution, but I can't figure out how to create a new image from a subset of pixels from the original image. I feel like for this I could access the raw pixel data using `ImageExt::to_raw_data()` and create a new `u8` array with the pixels I need, but I'm unsure how to then take that new array and turn it into a new image, as well as how to deal with different color formats to figure out how many `u8`s in the array correspond to a single pixel.
3) Spritesheet reuse: The spritesheet is loaded once into memory, and each item in the sprite list renders a scaled reference to the original sprite sheet such that the marked sprite is all that fits within the frame for that particular sprite. I can get the image to scale and offset to fit any desired rectangle within a frame, but I can't figure out how to reuse the same image, as setting a frame's image seems to take ownershup of the image struct and I can't then use it as the image in additional frames elsewhere in the program.

Knowing how to reuse images would also be useful when creating UI that might need custom icons to be reused in a few places.

Thanks in advance for any advice

**Comment by andrea-gamedev** [comment]
Update: I found the `RgbImage` type and managed to implement idea 2 using `ImageExt::to_rgb_data` and `RgbImage::new`. It gets a bit laggy with large numbers of sprites, but I have some ideas for optimizations I could make for speed, and the memory footprint is acceptable, except maybe for substantially larger projects with more spritesheets than what I'm working with. However, if option 1 is possible using fltk-rs, I'd still love to know, as that would bring the memory footprint down from acceptable to basically negligable even for worst case use cases for the application I'm trying to build

**Comment by MoAlyousef** [comment, marked as answer]
You can use the ImageExt::draw_ext method which takes an x_offset and y_offset:
```rust
use fltk::{prelude::*, *};
use std::cell::RefCell;
use std::rc::Rc;

const SPRITE_WIDTH: i32 = 172;
const SRPITE_HEIGHT: i32 = 183;

fn main() {
    let a = app::App::default();
    let mut wind = window::Window::default().with_size(400, 300);
    let mut frame = frame::Frame::new(50, 50, SPRITE_WIDTH, SRPITE_HEIGHT, None);
    let mut btn = button::Button::default()
        .with_size(SPRITE_WIDTH, 30)
        .below_of(&frame, 5)
        .with_label("Move");
    wind.end();
    wind.make_resizable(true);
    wind.show();

	let png = image::PngImage::load("spritesheet.png").unwrap();
    let mut rgb = png.to_rgb().unwrap();
	let factor_rc = Rc::from(RefCell::from(0));
    let factor = factor_rc.clone();
    frame.draw(move |f| {
        rgb.draw_ext(
            f.x(),
            f.y(),
            SPRITE_WIDTH,
            SRPITE_HEIGHT,
            *factor.borrow(),
            0,
        );
    });
    btn.set_callback(move |_| {
        *factor_rc.borrow_mut() += SPRITE_WIDTH;
        frame.parent().unwrap().redraw();
    });
	
    a.run().unwrap();
}
```
For this spritesheet.png for example:
![spritesheet](https://user-images.githubusercontent.com/37966791/137212867-7788c972-565f-405c-83fe-7ca96943c14b.png)

![image](https://user-images.githubusercontent.com/37966791/137213044-e90a77cb-62a6-4ccf-bb0c-f4e95fd805b0.png)

![image](https://user-images.githubusercontent.com/37966791/137213096-ea223054-c469-4acb-99ee-f3b2adf9c3cd.png)

**Reply by andrea-gamedev** [reply]
What if I wanted to draw different frames from the same spritesheet in multiple different widgets, and have a scrolling list of each sprite in the sheet? It won't let me since once I move the `RgbImage` into the closure, ownership is transferred to the closure and the `RgbImage` can't be used in the next iteration of the loop that creates the list items

```rust
let img_data = fltk::image::PngImage::load("exterior_tileset.png").expect("Failed to load image");
let mut img = img_data.to_rgb().expect("Failed to convert image to generic type");

let sprites_w = img.w() / SPRITE_SIZE;

let scroll = fltk::group::Scroll::default().size_of_parent();

for i in 0..num_sprites {
	let mut sprite_frame = fltk::frame::Frame::default()
		.with_size(SPRITE_SIZE, SPRITE_SIZE)
		.with_pos(0, i as i32 * SPRITE_SIZE);
	sprite_frame.set_frame(fltk::enums::FrameType::UpBox);
	sprite_frame.draw(move |frame| {
		let sprite_x = i as i32 % sprites_w;
		let sprite_y = i as i32 / sprites_w;
		img.draw_ext(
			frame.x(), frame.y(),
			SPRITE_SIZE, SPRITE_SIZE,
			sprite_x * SPRITE_SIZE, sprite_y * SPRITE_SIZE,
		);
	});
}

scroll.end();
```

```
error[E0382]: use of moved value: `img`
  --> src/main.rs:35:21
   |
24 |     let mut img = img_data.to_rgb().expect("Failed to convert image to generic type");
   |         ------- move occurs because `img` has type `RgbImage`, which does not implement the `Copy` trait
...
35 |         sprite_frame.draw(move |frame| {
   |                           ^^^^^^^^^^^^ value moved into closure here, in previous iteration of loop
...
38 |             img.draw_ext(
   |             --- use occurs due to use in closure
```

This is some test code not meant for the application, I'm just taking a spritesheet that consists of nothing but 16x16 sprites and trying to draw a list with each of them, left to right, top to bottom, using the largest spritesheet from my game dev project to stress test my solution. The frame type is just there so I could make sure all the boxes were placed correctly and scrolled correctly before adding the image.

**Reply by MoAlyousef** [reply]
Cloning the RgbImage is a cheap operation, it doesn’t create a new image, it’s only a shallow copy which copies the address of the image. To get a deep copy, you’ll have to use the copy() function.

**Reply by andrea-gamedev** [reply]
Damn I completely forgot, this all works and with less than half the memory footprint of my option 2 solution and no discernible lag or framerate loss with over 6000 entries in the list

I'm still relatively new to rust and I'm still forgetting the difference between `copy` and `clone`

Thanks for the help!

**Reply by andrea-gamedev** [reply]
Add-on question: If I wanted to draw a 16x16 sprite but have it fill a 48x48 frame so that the "preview" of the sprite isn't tiny, how would I go about scaling that? I've tried calling `ImageExt::scale` in different places and with different combinations of arguments but nothing has worked so far, I either just get the same 16x16 sprites in the upper left corner of an mostly empty 48x48 frame, or I get nothing

**Reply by MoAlyousef** [reply]
You can try frame.set_image_scaled(). Or if you’re drawing the image in the draw method, try scaling then drawing. Unfortunately I don’t have a pc at hand to check things out.

**Reply by andrea-gamedev** [reply]
I'm not really sure how to use `set_image_scaled` in this context, since I'm drawing in the `draw` method using `draw_ext`. I tried it inside and outside the `draw` method, before and after `draw_ext` and got the results I described above. I also tried calling `scale` both inside and outside the `draw` method, and before and after `draw_ext` and got the same thing.

**Reply by MoAlyousef** [reply]
In the draw method, you use ImageExt::draw directly:
```rust

frame.draw(move |f| {
        rgb.draw(
            f.x(),
            f.y(),
            f.w(),
            f.h(),
        );
    });

```

Normally this should draw the image to the size of the widget.

**Reply by JamesBear** [reply]
Is there a way to get size of a just loaded SharedImage?

**Reply by MoAlyousef** [reply]
You can call the w() and h() methods.

**Reply by andrea-gamedev** [reply]
~~Using `ImageExt::draw` does seem to do that, but then I can't use the offset that I need for drawing different sprites found throughout the image, I can only draw the top left sprite in the spritesheet~~

Never mind, I just needed to scale the source image before cloning it. I swear I tried that before and it gave me weird overflow errors when compiling, but it works now so...

```rust
let scale = LIST_ITEM_SIZE / SPRITE_SIZE;
img.scale(img.w() * scale, img.h() * scale, true, true);

for i in 0..num_sprites {
	let mut img_clone = img.clone();

	let mut sprite_frame = fltk::frame::Frame::default()
		.with_size(LIST_ITEM_SIZE, LIST_ITEM_SIZE)
		.with_pos(0, i as i32 * LIST_ITEM_SIZE);
	sprite_frame.set_frame(fltk::enums::FrameType::UpBox);
	sprite_frame.draw(move |frame| {
		let sprite_x = i as i32 % sprites_w;
		let sprite_y = i as i32 / sprites_w;
		img_clone.draw_ext(
			frame.x(), frame.y(),
			LIST_ITEM_SIZE, LIST_ITEM_SIZE,
			sprite_x * SPRITE_SIZE * scale, sprite_y * SPRITE_SIZE * scale,
		);
	});
}
```

<img width="412" alt="image" src="https://user-images.githubusercontent.com/53105809/137496973-b942f92b-a5da-4965-b499-f2408dc40ac9.png">

**Reply by MoAlyousef** [reply]
Great to hear

**Reply by andrea-gamedev** [reply]
I think you might have misunderstood my question but I got it working and edited my comment above with the solution.

The last thing I'd like to know is how to make it scale while preserving hard edges between pixels rather than blurring them. The `FL_Shared_Image` class in FLTK has a method `scaling_algorithm` that lets you choose between nearest and bilinear scaling, but I can't find an equivalent in fltk-rs

**Reply by MoAlyousef** [reply]
That function is currently not exposed in the Rust bindings. I can expose it in the next release.

**Reply by andrea-gamedev** [reply]
Thank you for all the help and fast responses

What is your policy on public contributions by the way? I wouldn't mind taking a crack at adding it myself, or if you would rather do that I'd love to contribute to the documentation and add more explanations, code examples, screenshots, etc

**Reply by MoAlyousef** [reply]
Contributions are welcome.

Usually contributing to the raw bindings requires also a contribution to cfltk first then to fltk-rs. 
There’s a [CONTRIBUTING.md](https://github.com/fltk-rs/fltk-rs/blob/master/CONTRIBUTING.md) which tries to explain the structure of fltk-rs.
